### ISSUE TITLE
Question about the implementation of 'EdgePrediction'

### ISSUE THREAD

**Issue by AlexHex7** [DeepGraphLearning/torchdrug#107]
Hi,  I think somethimes the **neg_index pairs** will be the same as **node_in&node_out pairs**. However, the first 50% part samples (node_in%node_out) are positive labels. And the others are negative labels (neg_index) (.i.e **self.target**). Is that wrong? or anything I misunderstand?

~~~
def predict(self, batch, all_loss=None, metric=None):
    graph = batch["graph"]

    output = self.model(graph, graph.node_feature.float(), all_loss, metric)
    node_feature = output["node_feature"]

    graph = self._get_directed(graph)
    node_in, node_out = graph.edge_list.t()[:2]

    # ============== The Code I Talk About =========================
    neg_index = (torch.rand(2, graph.num_edge, device=self.device) * graph.num_nodes[graph.edge2graph]).long()
    neg_index = neg_index + (graph.num_cum_nodes - graph.num_nodes)[graph.edge2graph]
    # ========================================================
    node_in = torch.cat([node_in, neg_index[0]])
    node_out = torch.cat([node_out, neg_index[1]])

    pred = torch.einsum("bd, bd -> b", node_feature[node_in], node_feature[node_out])
    return pred

# ============== The Code I Talk About =========================
def target(self, batch):
    graph = batch["graph"]
    target = torch.ones(graph.num_edge, device=self.device)
    target[graph.num_edge // 2:] = 0
    return target
# ========================================================

def forward(self, batch):
    """"""
    all_loss = torch.tensor(0, dtype=torch.float32, device=self.device)
    metric = {}

    # ============== The Code I Talk About =========================
    pred = self.predict(batch, all_loss, metric)
    target = self.target(batch)
    loss = F.binary_cross_entropy_with_logits(pred, target)
    # ========================================================

    name = tasks._get_criterion_name("bce")
    metric[name] = loss
    metric.update(self.evaluate(pred, target))

    all_loss += loss

    return all_loss, metric
~~~

**Comment by KiddoZhu** [comment]
You're right. The negative samples for edge prediction are not implemented to be strict negatives. For each batch, the first half are positive samples, and the second half are (possibly) negative samples.

**Comment by AlexHex7** [comment]
Got it.  Thanks for your reply! @KiddoZhu
